dbms/src/Storages/Page/V3/PageDirectoryFactory.cpp

```diff
@@ -42,6 +42,22 @@ PageDirectoryPtr PageDirectoryFactory::create(String storage_name, FileProviderP
     dir->gcInMemEntries();
     LOG_FMT_INFO(DB::Logger::get("PageDirectoryFactory"), "PageDirectory restored [max_page_id={}] [max_applied_ver={}]", dir->getMaxId(), dir->sequence);
 
+    // MVCC gc must do twice when restore
+    // Because the first time call MVCC GC may not clean the deleted entries if being reference.
+    // But it will clean up the deleted reference.
+    // ex.
+    //  P 1
+    //  R 2 -> 1
+    //  D 1
+    //  D 2
+    // First time call gcInMemEntries():
+    //  - P 1 entry 0 being_ref_count is not zero, won't be cleaned.
+    //  - R 2 entry 0 being_ref_count is zero, deref the P 1, make it being_ref_count is 0. Also R2 will be clean up.
+    //
+    // Second time call gcInMemEntries():
+    //  - P 1 being_ref_count is 0, it will be clean.
+    dir->gcInMemEntries();
```

Comment:
Also add it to `PageDirectoryFactory::createFromEdit`

Reply:
done

Reply:
I mean in these lines, we should also call gcInMemEntries twice
https://github.com/pingcap/tiflash/blob/62e7c14141c0a3390ac5a4454708126511382838/dbms/src/Storages/Page/V3/PageDirectoryFactory.cpp#L87-L100